File: study/8주차 문제/최혜령/택배 배달과 수거하기.py
def solution(cap, n, deliveries, pickups):
    answer = 0  # 왕복 거리의 총합을 저장할 변수
    deliverySum = 0  # 배달할 물품의 누적량
    pickupSum = 0    # 수거할 물품의 누적량
    
    # n-1번째 집부터 0번째 집까지 반복 (뒤에서부터 처리)
    for i in range(n-1, -1, -1):
        cnt = 0  # 왕복 횟수를 세기 위한 변수
        deliverySum += deliveries[i]  # 현재 집의 배달할 물품을 누적
        pickupSum += pickups[i]       # 현재 집의 수거할 물품을 누적
        
        # 배달할 물품이나 수거할 물품이 남아 있는 동안 반복
        while deliverySum > 0 or pickupSum > 0:
            deliverySum -= cap  # 배달할 수 있는 최대 용량만큼 배달
            pickupSum -= cap    # 수거할 수 있는 최대 용량만큼 수거
            cnt += 1  # 한 번 왕복했으므로 카운트를 증가시킴
        
        # 해당 집까지의 왕복 거리를 총합에 더함
        answer += (i + 1) * 2 * cnt  # (i+1)은 해당 집까지의 거리, 2는 왕복 거리, cnt는 왕복 횟수
    
    return answer  # 최종 계산된 총 왕복 거리를 반환

File: study/8주차 문제/장원석/택배배달과수거하기.java
import java.util.ArrayDeque;
import java.util.Deque;

public class 택배배달과수거하기 {

	public static void main(String[] args) {
		System.out.println(solution(4, 5, new int[] { 1, 0, 3, 1, 2 }, new int[] { 0, 3, 0, 4, 0 }));
		System.out.println(solution(2, 7, new int[] { 1, 0, 2, 0, 1, 0, 2 }, new int[] { 0, 2, 0, 1, 0, 2, 0 }));
	}

	private static long solution(int cap, int n, int[] deliveries, int[] pickups) {
		long answer = 0;

		Deque<int[]> queue1 = new ArrayDeque<>();
		Deque<int[]> queue2 = new ArrayDeque<>();

		for (int i = 0; i < n; i++) {
			if (deliveries[i] != 0)
				queue1.addLast(new int[] { deliveries[i], i + 1 });

			if (pickups[i] != 0)
				queue2.addLast(new int[] { pickups[i], i + 1 });
		}

		while (!queue1.isEmpty() || !queue2.isEmpty()) {
			int sum = 0;
			int dist = 0;

			while (!queue1.isEmpty() && sum < cap) {
				int[] delivery = queue1.pollLast();
				dist = Math.max(dist, delivery[1]);

				if (delivery[0] + sum <= cap) {
					sum += delivery[0];
				} else {
					delivery[0] -= (cap - sum);
					queue1.addLast(delivery);
					break;
				}
			}

			sum = 0;

			while (!queue2.isEmpty() && sum < cap) {
				int[] pickup = queue2.pollLast();
				dist = Math.max(dist, pickup[1]);

				if (pickup[0] + sum <= cap) {
					sum += pickup[0];
				} else {
					pickup[0] -= (cap - sum);
					queue2.addLast(pickup);
					break;
				}
			}

			answer += dist * 2;
		}

		return answer;
	}
}
